cid-decode/src/multicodecs.rs

```diff
@@ -0,0 +1,617 @@
+use failure::err_msg;
+use ipfs_resolver_common::Result;
+use std::collections::HashMap;
+
+pub struct Multicodec {
+    name: &'static str,
+    pub tag: Tag,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum Tag {
+    Cid,
+    Filecoin,
+    Holochain,
+    Ipld,
+    Key,
+    Libp2p,
+    Multiaddr,
+    Multiformat,
+    Multihash,
+    Namespace,
+    Serialization,
+    Softhash,
+    Transport,
+    Zeroxcert,
+    Unkown,
+}
+
+impl Multicodec {
+    fn new(name: &'static str, tag: &str) -> Self {
+        Self {
+            name,
+            tag: match tag {
+                "cid" => Tag::Cid,
+                "filecoin" => Tag::Filecoin,
+                "holochain" => Tag::Holochain,
+                "ipld" => Tag::Ipld,
+                "key" => Tag::Key,
+                "libp2p" => Tag::Libp2p,
+                "multiaddr" => Tag::Multiaddr,
+                "multiformat" => Tag::Multiformat,
+                "multihash" => Tag::Multihash,
+                "namespace" => Tag::Namespace,
+                "serialization" => Tag::Serialization,
+                "softhash" => Tag::Softhash,
+                "transport" => Tag::Transport,
+                "zeroxcert" => Tag::Zeroxcert,
+                _ => Tag::Unkown,
+            },
+        }
+    }
+
+    pub fn check_tag_get_name(&self, tag: Tag) -> Result<&'static str> {
+        if tag != self.tag {
+            return Err(err_msg(format!(
+                "Multicodec {} has type {:?} expected to have type {:?}",
+                self.name, self.tag, tag
+            )));
+        }
+        Ok(self.name)
+    }
+
+    pub fn log_unknown_multicodec(&self, line_no: usize, line: &str) -> &Self {
+        if UNKNOWN_MULTICODEC.tag == self.tag {
+            warn!("Unknown Multicodec found (row {}: {})", line_no, line)
+        }
+        self
+    }
+}
+
+pub struct MulticodecTable {
+    table: &'static HashMap<u64, Multicodec>,
+}
+
+impl MulticodecTable {
+    pub fn get(&self, codec: u64) -> &'static Multicodec {
```

Comment:
Alright, this is where the meat is at: I'd change this to something like this:
```rust
pub fn get_checking_tag(&self, code: u64, tag: Tag) -> Option<Result<&'static Multicodec>> {
  // If not present in table: return None
  // If present but wrong tag: Return Some(Err(some_clever_message))
  // If present and correct tag: return Some(Ok(mc))
}
```
(To implement that, you probably want something like [`Option::map`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) and [`Option::ok_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or) or [`Option::ok_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) (which is more efficient, since it only computes the `else` if the option is None))

That transforms the invocation to something like this:
```rust
multicodecs::MULTICODEC_TABLE
  .get_checking_tag(c.codec(), Tag::Ipld)
  .ok_or(err_msg("unknown multicodec"))? // Unpack the Option, handling the not-present-in-table case
  .context("non-IPLD multicodec")? // Unpack the result, handling the invalid-tag case
```

And then make this the only way to access the table and get a Multicodec from it. That implicitly always checks the tag, which ensures we can't ever forget to do that.

What do you think?

Reply:
I think this is really the best way to do it! Will implement!

Reply:
Done.